### ISSUE TITLE
Add support for local links in PDF files

### ISSUE THREAD

**Issue by DVK251** [koreader/kindlepdfviewer#72]
I think it will be useful to add support for local links in PDF files. I don't mean a links which point to another pdf files or Internet URLs. I mean links which points to other pages of the original document. It is handy thing when reading books which contain endnotes.

**Comment by hwhw** [comment]
Agreed. And I also had a look. I was waiting for houqp to submit the djvu highlighting - I want to implement it for PDFs, too. When we have cursor handling, adding support for links is the next step. Actually it isn't very hard :-) And it seems I'll have some time to work on it over the weekend.

**Comment by hwhw** [comment]
So now we have highlighting support even for PDFs. I'd like to wait with the local links until after release, though.

**Comment by eLiNK2gl** [comment]
After the shift + 'X' mode, this feature might be another enrichment, which would make kpdfview stand head and shoulders above alternatives.

Any chance to see it implemented in the short-run?

**Comment by tigran123** [comment]
@eLiNK2gl I would be very happy with it as almost all the books I typeset use this feature (e.g. all the page numbers in my TOC are hyperlinks to the actual pages and cross-references jump to the appropriate location within a book). Interesting, @hwhw says it is not very hard... Maybe I'll have a look at some point, but I don't know anything about muPDF internals --- I only know djvulibre :)

**Comment by dpavlin** [comment]
I just pushed first naive implementation of mupdf bindings to get links on pdf page.

I'm not submitting pull request because there is no user-visible functionality (yet), but I would appreciate if you can take a look at code to make sure that I didn't do something wrong.

**Comment by tigran123** [comment]
Very interesting, thank you. I will test it this evening and will let you know how it went.

**Comment by tigran123** [comment]
I couldn't resist testing it quickly in the emulator --- yes, it works very nicely! I tested it on the pdf file of the Urantia Study Edition (see "urantia-study-edition" project on github). The TOC there contains links (visually unnoticeable to not distract attention).

**Comment by tigran123** [comment]
Btw, I only tested that all the links are present, but did not verify the actual page destinations. In fact, this should be verified in two cases: a) simple arabic numerals from the beginning and b) roman numerals for front matter and then arabic for main matter, i.e. so-called "logical page numbering". In all of my recent editions I switched to simple arabic numerals from the very beginning (btw, in Soviet Union this was a required standard and even as recently as about 10 years ago I published a hardcover Greek NT in Ukraine which had the same strict requirements on page numbering) but I should still have some of my old editions with roman+arabic numerals to test on.
